lib/Tools/LinearLayout.cpp

```diff
@@ -0,0 +1,382 @@
+          Twine(outer.toString()) + "\nInner:\n" + inner.toString());
+    }
+  };
+
+  // Check that dims common to outer and inner have the same relative order.
+  checkCommonDims(outer.getInDimNames(), inner.getInDimNames());
+  checkCommonDims(outer.getOutDimNames(), inner.getOutDimNames());
+
+  // Get the sizeLog2 of all inner and outer dimensions we're going to consider,
+  // in order.  `inner` is more minor, so its dimensions come first.
+  llvm::MapVector<StringAttr, int32_t> inDimSizes;
+  llvm::MapVector<StringAttr, int32_t> outDimSizes;
+  for (const auto &layout : {inner, outer}) {
+    for (StringAttr inDim : layout.getInDimNames()) {
+      inDimSizes[inDim] += layout.getInDimSizeLog2(inDim);
+    }
+    for (StringAttr outDim : layout.getOutDimNames()) {
+      outDimSizes[outDim] += layout.getOutDimSizeLog2(outDim);
+    }
+  }
+
+  BasesT allBases;
+  for (auto [inDimName, inDimSize] : inDimSizes) {
+    for (auto [outDimName, outDimSize] : outDimSizes) {
+      std::vector<int32_t> bases(inDimSize);
+      if (inner.hasInDim(inDimName) && inner.hasOutDim(outDimName)) {
+        for (int i = 0; i < inner.getInDimSizeLog2(inDimName); i++) {
+          bases[i] = inner.getBasis(inDimName, outDimName, i);
+        }
+      }
+      if (outer.hasInDim(inDimName) && outer.hasOutDim(outDimName)) {
+        int offset =
+            inner.hasInDim(inDimName) ? inner.getInDimSizeLog2(inDimName) : 0;
+        int shift = inner.hasOutDim(outDimName)
+                        ? inner.getOutDimSizeLog2(outDimName)
+                        : 0;
+        for (int i = 0; i < outer.getInDimSizeLog2(inDimName); i++) {
+          bases[offset + i] = outer.getBasis(inDimName, outDimName, i) << shift;
+        }
+      }
+      allBases[inDimName][outDimName] = std::move(bases);
+    }
+  }
+
+  return LinearLayout(std::move(allBases));
+}
+
+SmallVector<std::pair<StringAttr, int32_t>>
+LinearLayout::apply(ArrayRef<std::pair<StringAttr, int32_t>> ins) const {
+  // `ins` must contain the same dimensions as getInDimNames(), modulo
+  // reordering.
+  auto insNames = llvm::make_first_range(ins);
+  llvm::DenseSet<StringAttr> argInDims(insNames.begin(), insNames.end());
+  llvm::DenseSet<StringAttr> thisInDims(getInDimNames().begin(),
+                                        getInDimNames().end());
+  if (argInDims != thisInDims) {
+    llvm::report_fatal_error(
+        Twine("Cannot apply layout.  The given input dimensions must match the "
+              "layout's input dimensions, modulo reordering, but they don't.") +
+        "\nGiven dims:" + join(insNames, ", ") +
+        "\nLayout's dims:" + join(getInDimNames(), ", "));
+  }
+
+  SmallVector<std::pair<StringAttr, int32_t>> ret;
+  for (StringAttr outDim : getOutDimNames()) {
+    int32_t outVal = 0;
+    for (auto &[inDim, val] : ins) {
+      for (int i = 0; i < getInDimSizeLog2(inDim); i++) {
+        if (val & (1 << i))
+          outVal ^= getBasis(inDim, outDim, i);
+      }
+    }
```

Comment:
> all these comments apply verbatim to applyLinearLayout, where the performance is important

LLVM will optimize it.  We very much rely on this in either case, because applying LLs bit-by-bit is not optimal.

(Also I'm not sure what you're suggesting we change in the emitted code, because you can't do `while (v) { ... v >> 1; }` in the emitted code if v is a value.)

Reply:
in the emitted code you would do `for _ in range(log2(v)):`